bpf/lib/policy_log.h

```diff
@@ -28,11 +28,23 @@ struct policy_verdict_notify {
 	__u32	pad1; // align with 64 bits
 };
 
+static __always_inline bool policy_verdict_filter_allow(__u32 filter, __u8 dir)
+{
+	// Doesn't work somehow on my setup
+	// return ((filter & (1 << (dir-1))) > 0);
+	// return ((filter & dir) > 0);
+	return ((filter & (0x40000000 + dir)) > 0);
```

Comment:
Does this need a follow up?

Reply:
It would be nice if someone from Cilium can help. It will also be very useful if we can have a way to load a bpf program and easily test the functionality on different kernel versions.

Reply:
Are you hitting issues with the current version in this PR?

Which kernel and compiler are you using when you hit these issues?

I would love to have a way to load & test a BPF program against different kernel verifier versions easily, but there's a lot of combinations and that code is unfortunately not particularly easy to rip out and run standalone.

Overall our current CI testing is based on validating that with LLVM-7 and either a recent 4.9.x kernel version or a recent bpf-next / net-next kernel prerelease version, and if those two work, then we can pass through to the next stage which typically involves developers and users running the code on a wider variety of kernels. If this code is not performing correctly on a particular kernel, that may be a kernel bug; in that case, we need to identify whether the bug was already fixed somewhere (eg by testing other kernel versions) and whether the relevant fix was backported by either upstream maintainers or distro maintainers.

Reply:
"Are you hitting issues with the current version in this PR?" Do you mean current code? Current code works. The two commented way didn't work for me. I hit the issue for both 4.14 and 4.19 kernel.

Reply:
I tried the two commented lines on a 4.19 kernel with Clang 7 and all policy verdicts are filtered regardless of network policies. If I try the third line (i.e., this pull request), it works as expected, I see policy verdicts according to ingress/egress network policies. @lzang Is that the issue you were having as well?

Reply:
Yes, that is the issue I am having. Thanks for trying it out!

Reply:
Ok. So I added a function in its own section in `bpf_lxc.c` with only a call to `send_policy_verdict_notify()` to be able to look at the generated bytecode for the diverse C codes above. With the third line (the one that works), I get:
```
       0:	bf 16 00 00 00 00 00 00	r6 = r1
       1:	18 01 00 00 01 00 00 80 00 00 00 00 00 00 00 00	r1 = 2147483649 ll
       3:	18 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00	r2 = 0 ll
       5:	5f 12 00 00 00 00 00 00	r2 &= r1
       6:	15 02 1f 00 00 00 00 00	if r2 == 0 goto +31 <LBB14_4>
[...]
0000000000000130 LBB14_4:
      38:	b7 00 00 00 00 00 00 00	r0 = 0
      39:	95 00 00 00 00 00 00 00	exit
```
Whereas, with the second line, I get:
```
       0:	b7 00 00 00 00 00 00 00	r0 = 0
       1:	95 00 00 00 00 00 00 00	exit
```
So the compiler seem to think that, because `filter` is 0 at compile time, it can optimize out this logic and get rid of most of the `send_policy_verdict_notify` function. It's not yet clear to me why the compiler think it's okay to optimize or why the third line doesn't get optimized in a similar fashion.

In any case, we may want to use `volatile` here instead and add a comment to explain.
```c
volatile __u8 d = dir;
return ((filter & d) > 0);
```
Both Clang 7 and 9 behave this way. I haven't checked newer versions yet.

Reply:
@pchaigno Thanks for your investigation. We've already made the filter to be always none-zero as we knew initializing it to zero causes issue. Filter is a volatile uint32 and its value is correctly interpreted. The issue here is dir, which is an input to the function and is always non-zero. Making a local variable to be volatile looks strange to me. Did you try this change end-to-end?

Reply:
`filter` is actually zero when compiling (`r2` in my above snippet) and is replaced with an actual value when we rewrite the object file. [This LPC presentation](http://vger.kernel.org/lpc_bpf2018_talks/bpf_global_data_and_static_keys.pdf) gives some background and explanations for those config. variables. Basically, `POLICY_VERDICT_LOG_FILTER` ends up taking the address of the global variable and not its value; that allows us to easily rewrite it in the object file afterwards. In BPF, such addresses of global variables are usually references to maps and rewritten in the kernel at load time to point to the actual BPF map. *My guess* is that the compiler sets it to zero because it assumes it's one of this map cases where the compile-time value doesn't matter.

I did test the change I proposed above. I executed Cilium with the same setup as before (Clang 7) and checked that policy verdicts were filtered by default and started showing up when network policies were added. I did not run the whole test suite though.

Reply:
@pchaigno What I don't understand is that you are talking about filter being zero or not, but your fix is on dir. Making dir to be volatile or adding some constant to it are both trying to change dir.

Reply:
You're right, that wasn't super clear. There are two different things here:
1. Why the compiler thinks the condition is always equivalent to zero. That's because `filter` is always zero at compile time and that's what I tried to explain in my previous comment.
0. Why making `dir` volatile prevents the code optimization. I don't fully understand that one. I would expect that having one variable volatile in the expression (i.e., `filter`) is enough, but that doesn't seem to be the case. Making both variables volatile seems to be what the compiler needs to understand that the statement cannot be optimized out. I tested this with Clang 7 and Clang trunk (11) and they both behave this way.